File: Python/odin/group.py
from psycopg2.extras import Json


SET_GROUP = '''INSERT INTO odin.group_ledger
    (reference, group_slug, description)
    VALUES (%s, %s, %s)'''
SET_MEMBERSHIP = '''INSERT INTO odin.group_membership_ledger
    (reference, identity_id, group_slug, member)
    VALUES (%s, %s, %s, %s)'''
SET_PERMISSION = '''INSERT INTO odin.group_grant_ledger
    (reference, group_slug, permission_slug, allows)
    VALUES (%s, %s, %s, %s)'''


def setgroup(cnx, group, description=''):
    cnx.assert_module('authz')
    cnx.execute(SET_GROUP, (cnx.reference, group, description))
    print(group, "set up")


def addmembership(cnx, user, *groups):
    cnx.assert_module('authz')
    for group in groups:
        cnx.execute(SET_MEMBERSHIP, (cnx.reference, user, group, True))
        print(user, "is a member of", group)


def assignpermission(cnx, group, *permissions):
    cnx.assert_module('authz')
    for permission in permissions:
        cnx.execute(SET_PERMISSION, (cnx.reference, group, permission, True))
        print(group, "now allows", permission)

File: Cpp/odin/group.cpp
/**
    Copyright 2018 Felspar Co Ltd. <http://odin.felspar.com/>

    Distributed under the Boost Software License, Version 1.0.
    See <http://www.boost.org/LICENSE_1_0.txt>
*/

#include <fostgres/sql.hpp>

#include <odin/group.hpp>
#include <odin/fg/native.hpp>
#include <odin/pwhashproc.hpp>

#include <fost/insert>


void odin::group::alter_membership (
    fostlib::pg::connection &cnx,
    f5::u8view reference,
    f5::u8view identity_id,
    f5::u8view group_slug,
    bool is_member
) {
    fg::json group_membership;
    fostlib::insert(group_membership, "reference", reference);
    fostlib::insert(group_membership, "identity_id", identity_id);
    fostlib::insert(group_membership, "group_slug", group_slug);
    fostlib::insert(group_membership, "member", is_member);
    cnx.insert("odin.group_membership_ledger", group_membership);
}
